test/base/populators.py

```diff
@@ -858,9 +872,15 @@ def wait_on(function, desc, timeout=DEFAULT_TIMEOUT):
             timeout_message = "Timed out after %s seconds waiting on %s." % (
                 total_wait, desc
             )
-            assert False, timeout_message
+            raise TimeoutAssertionError(timeout_message)
```

Comment:
If the last `value` returned by `function()` was also an attribute of `TimeoutAssertionError`, you could skip the calls at lines 202 and 809. I'm not particularly concerned about the extra time spent, but these calls could actually return a different value than the last.

Reply:
I thought about something like this but it isn't quite right is it? ``wait_on`` isn't being sent ``function``s with useful return values. 

- For 202, this is running ``has_active_jobs`` and the error message uses ``get_jobs`` - the last return of ``has_active_jobs`` isn't useful right (it is just going to be ``True``)? I considered using a ``local`` declaration in there but it seemed more awkward than this.
- For 809, this is using ``get_state`` but the error message uses ``state_func``.

If we had Python 3 ``nonlocal`` we could save the more useful result in the function - but it is a bit more awkward in Python 2. I was aware this isn't quite returning the last value from the API used in the waiting but I used language like "current" instead of "last" as a result.

Reply:
That's fair enough, thanks for the thorough explanation!